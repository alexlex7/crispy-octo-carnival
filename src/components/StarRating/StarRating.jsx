import React, { useMemo, useState } from 'react';
import { ReactComponent as StarIcon } from '../../assets/images/star.svg';
import styles from './StarRating.module.css';
import throttle from 'lodash.throttle';
import PropTypes from 'prop-types';

export default function StarRating({
  handleSetRating = () => {},
  rating,
  isReadOnly,
}) {
  const [hoverRating, setHoverRating] = useState(0);
  const [isMouseOver, setIsMouseOver] = useState(false);

  const choseRating =
    isMouseOver && !isReadOnly ? hoverRating : (rating / 5) * 200;

  const starsActiveWrapperWidth = (rating) => {
    const width = (rating / 200).toPrecision(1) * 100;
    let countWidth = 0;
    if (width < 10) {
      countWidth = width < 7 ? 0 : 10;
    } else if ((width / 10) % 2 === 1) {
      countWidth = width;
    } else {
      countWidth = width;
    }
    return `${countWidth}%`;
  };

  const mouseOverHandler = useMemo(() => {
    const throttled = throttle((e) => {
      const { layerX } = e.nativeEvent;
      setHoverRating(layerX);
    }, 100);
    return (e) => {
      e.persist();
      return throttled(e);
    };
  }, []);

  return (
    <div className={styles.ratingWrapper}>
      <div
        className={styles.starsWrapper}
        onMouseOver={mouseOverHandler}
        onClick={(e) => {
          handleSetRating(e.nativeEvent.layerX);
        }}
        onMouseEnter={() => {
          setIsMouseOver(true);
        }}
        onMouseLeave={() => {
          setIsMouseOver(false);
        }}
      >
        {[1, 2, 3, 4, 5].map((index) => (
          <StarIcon key={index} className={styles.icon} fill="none" />
        ))}
      </div>
      <div
        className={styles.starsActiveWrapper}
        style={{
          width: starsActiveWrapperWidth(choseRating),
        }}
      >
        <div style={{ display: 'flex' }}>
          {[1, 2, 3, 4, 5].map((index) => (
            <StarIcon key={index} className={styles.activeIcon} fill="orange" />
          ))}
        </div>
      </div>
    </div>
  );
}

StarRating.propTypes = {
  handleSetRating: PropTypes.func,
  rating: PropTypes.number.isRequired,
  isReadOnly: PropTypes.bool,
};
