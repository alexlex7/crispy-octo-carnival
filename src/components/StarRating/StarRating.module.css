.ratingWrapper {
  position: relative;
  width: 200px;
}
.starsWrapper {
  display: flex;
}

.starsActiveWrapper {
  position: absolute;
  top: 0;
  width: 200px;

  display: flex;
  flex-wrap: nowrap;
  overflow: hidden;
  pointer-events: none;
}

.icon {
  display: block;
  width: 40px;
  height: 40px;
  fill: lightgray;
}
.activeIcon {
  display: block;
  width: 40px;
  height: 40px;
}
