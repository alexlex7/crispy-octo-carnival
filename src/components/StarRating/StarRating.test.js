import { render, screen } from '@testing-library/react';
import StarRating from './StarRating';

test('render star rating component', () => {
  const { asFragment } = render(<StarRating rating={4} />);
  expect(asFragment()).toMatchSnapshot();
});
