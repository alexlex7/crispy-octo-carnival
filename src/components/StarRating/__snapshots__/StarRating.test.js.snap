// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render star rating component 1`] = `
<DocumentFragment>
  <div
    class="ratingWrapper"
  >
    <div
      class="starsWrapper"
    >
      <svg
        class="icon"
        fill="none"
      >
        star.svg
      </svg>
      <svg
        class="icon"
        fill="none"
      >
        star.svg
      </svg>
      <svg
        class="icon"
        fill="none"
      >
        star.svg
      </svg>
      <svg
        class="icon"
        fill="none"
      >
        star.svg
      </svg>
      <svg
        class="icon"
        fill="none"
      >
        star.svg
      </svg>
    </div>
    <div
      class="starsActiveWrapper"
      style="width: 80%;"
    >
      <div
        style="display: flex;"
      >
        <svg
          class="activeIcon"
          fill="orange"
        >
          star.svg
        </svg>
        <svg
          class="activeIcon"
          fill="orange"
        >
          star.svg
        </svg>
        <svg
          class="activeIcon"
          fill="orange"
        >
          star.svg
        </svg>
        <svg
          class="activeIcon"
          fill="orange"
        >
          star.svg
        </svg>
        <svg
          class="activeIcon"
          fill="orange"
        >
          star.svg
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
